src/Command.ts

```diff
@@ -197,24 +197,103 @@ export default abstract class Command {
       return true;
     }
 
+    let inquirer: Inquirer | undefined;
+    const shouldPrompt = Cli.getInstance().getSettingWithDefaultValue<boolean>(settingsNames.prompt, false);
+
     const argsOptions: string[] = Object.keys(args.options);
-    for (const optionSet of optionsSets) {
-      if (optionSet.runsWhen === undefined || optionSet.runsWhen(args)) {
-        const commonOptions = argsOptions.filter(opt => optionSet.options.includes(opt));
+    const optionSetsWithoutArgs = optionsSets.filter(y => y.runsWhen === undefined);
+    const optionSetsWithArgs = optionsSets.filter(y => y.runsWhen);
+
+    for (const optionSet of optionSetsWithoutArgs) {
+      const commonOptions = argsOptions.filter(opt => optionSet.options.includes(opt));
+      if (commonOptions.length === 0) {
+        if (!shouldPrompt) {
+          return `Specify one of the following options: ${optionSet.options.join(', ')}.`;
+        }
```

Comment:
looks a bit like we duplicate a lot of code only to ensure we first go over normal required option sets and then those that are required when some condition is met 🤔.
What if we sort 🤔? Shouldn't something like this work
```typescript
private async validateOptionSets(args: CommandArgs, command: CommandInfo): Promise<string | boolean> {
    const optionsSets: OptionSet[] | undefined = command.command.optionSets;
    if (!optionsSets || optionsSets.length === 0) {
      return true;
    }

    let inquirer: Inquirer | undefined;
    const shouldPrompt = Cli.getInstance().getSettingWithDefaultValue<boolean>(settingsNames.prompt, false);
    const argsOptions: string[] = Object.keys(args.options);

    // ==== changed here ===
    // we foreach over all option sets but we sort them so first we always go over the 'normal' required and the 'runsWhen' go as second on the list ... remove this comment ;)
    for (const optionSet of optionsSets.sort(opt => opt.runsWhen ? 0 : 1)) {
      if (optionSet.runsWhen && !optionSet.runsWhen!(args)) { // we check if runsWhen needs to be checked .. also remove this comment ;)
        continue;
      }

    // ==== rest of the logic already implemented ===
      const commonOptions = argsOptions.filter(opt => optionSet.options.includes(opt));
      if (commonOptions.length === 0) {
        if (!shouldPrompt) {
          return `Specify one of the following options: ${optionSet.options.join(', ')}.`;
        }

        await this.promptForOptionSetNameAndValue(args, optionSet, inquirer);
      }

      if (commonOptions.length > 1) {
        if (!shouldPrompt) {
          return `Specify one of the following options: ${optionSet.options.join(', ')}, but not multiple.`;
        }

        await this.promptForSpecificOption(args, commonOptions, inquirer);
      }
    }

    return true;
  }
```

If I am not mistaken it should work  😉
> Famous last words 😅

What do you think? could you give it a check or probably I am missing some edge case here 🤔. But if this works we could remove a lot of duplicated lines 👍

Reply:
I have tested it with the same dummy-command as I did before and it seems to be working properly.

Reply:
> Thanks for the great first review and the kind words, @Adam-it.
> 
> Also, great suggestions. I have implemented them. Could you please have another look?

awesome I was of any help 🙂 
I will try to review the last changes ASAP. Sorry for the hold up 🙏